1.train_models/plate_5_model_training/plate_5_interpretable_models/nbconverted/plate_5_normalized_qc_optimize_xgb.py

```diff
@@ -0,0 +1,327 @@
+
+# In[9]:
+
+
+kwargs = {
+    "_welldf": welldf,
+    "_category_col": "Metadata_genotype",
+    "_well_col": "Metadata_Well",
+    "_cell_count_col": "Metadata_cell_count",
+    "_test_well_count": 3
+}
+
+
+# In[10]:
+
+
+gss = GreedySubsetSum()
+wells = gss.update_test_wells(**kwargs)
+
+
+# In[11]:
+
+
+restdf = platedf.loc[~platedf["Metadata_Well"].isin(wells)]
+testdf = platedf.loc[platedf["Metadata_Well"].isin(wells)]
+
+
+# ### Down-sample by genotype
+
+# In[12]:
+
+
+restdf = down_sample_by_genotype(restdf)
+testdf = down_sample_by_genotype(testdf)
+
+
+# # Process data
+
+# ## Encode labels
+
+# In[13]:
+
+
+le = LabelEncoder()
+y = le.fit_transform(restdf["Metadata_genotype"])
+
+
+# ## Filter columns
+
+# In[14]:
+
+
+X = restdf.drop(columns=meta_cols)
+X = X[feat_cols]
+X = filter_features(X)
+
+
+# ## XGboost-compatible data structure
+
+# In[15]:
+
+
+dX = xgb.DMatrix(X, label=y, nthread=-1)
+
+
+# # Train Models
+
+# ## Define parameters
+
+# In[16]:
+
+
+param_ranges = {
+    'learning_rate': (0.01, 0.51),
+    'subsample': (0.8, 1),
+    'colsample_bytree': (0.8, 1),
+    'gamma': (0, 0.4),
+    'reg_lambda': (0, 20),
+}
+
+
+# In[17]:
+
+
+xgb_params = {
+    "n_estimators": 5000,
+    "n_jobs": -1,
+    "random_state": 0,
+    "device": "cuda",
+    "verbosity": 0,
+    "importance_type": "total_gain",
+    "validate_parameters": True,
+    "objective": "multi:softmax",
+    "num_class": 3
+}
+
+
+# ## Randomly sample training hyperparameters
+
+# In[18]:
+
+
+rand_iter = 1400
+random_params = {
+    i:
+    {key: random.uniform(*param_ranges[key]) for key in param_ranges}
+    for i in range(rand_iter)
+}
+
+
+# ## Random Search
+
+# In[19]:
+
+
+for idx, rparams in random_params.items():
+
+    start_time = time.time()
+    comb_params = xgb_params | rparams
+    xgb_cv = xgb.cv(
+        params=comb_params,
+        dtrain=dX,
+        num_boost_round=100,
+        early_stopping_rounds=10,
+        nfold=5,
+        stratified=True,
+        seed=0,
+        metrics="merror",
+        verbose_eval=True
+    )
+    end_time = time.time()
+
+    min_merror_idx = xgb_cv["test-merror-mean"].argmin()
+
+    random_params[idx] = {
+        "min_test_merror_mean": xgb_cv["test-merror-mean"].iloc[min_merror_idx],
+        "min_test_merror_std": xgb_cv["test-merror-std"].iloc[min_merror_idx],
+        "min_test_merror_mean_idx": xgb_cv["test-merror-mean"].argmin()
+    }
+
+    print(f"Time to run iteration [[{idx}]] = ({(end_time - start_time) / 60:.2f} minutes)")
+
+
+# # Save model data
+
+# ## Format random search results
+
+# In[20]:
+
+
+rsrdf = pd.DataFrame.from_dict(random_params, orient='index')
+
+rsrdf.rename_axis("parameter_search_idx", inplace=True)
+
+rsrdf["max_test_accuracy_mean"] = 1 - rsrdf["min_test_merror_mean"]
+
+
+# ## Save random search results
+
+# In[21]:
+
+
+rsrdf.to_csv(f"{data_path}/random_search_xgb_opt_results.tsv", sep="\t")
+
+
+# ## Save split indices
```

Comment:
I'm a bit confused as to what this is? Why is doing this important? As is, it is also hard to determine if there are any explicit links to the split_indices file and the actual results - why do you need both? Is there a way to confirm alignment between indices and samples?